datafusion/physical-expr/src/aggregate/groups_accumulator/mod.rs

```diff
@@ -157,4 +161,14 @@ pub trait GroupsAccumulator: Send {
     /// in bytes. This function is called once per batch, so it should
     /// be `O(n)` to compute, not `O(num_groups)`
     fn size(&self) -> usize;
+
+    /// Try to reserve capacity the that at least a single [`RecordBatch`] can be inserted. The
```

Comment:
While implementing the reservation up front for all accumulators results in the most accurate memory accounting, it also adds significant complexity to implementing an accumulator. This API requires each implementation to attempt to 

Another approach, that DataFusion uses elsewhere is "optimistic" memory accounting. In that case, the used memory can exceed the target slightly (~ 1 RecordBatch) in order to simplify the accounting. 

If we were to take that approach with this PR, it would look something like:

```
Insert batch into GroupsAccumulator
call GroupsAccumulator::size()
  if exceeded limit, then spill
```

And then the existing GroupsAccumulators would not need to be changed

Reply:
Applied the "optimistic" approach and removed the upfront reservations.